src/excmds.ts

```diff
@@ -686,6 +686,21 @@ export async function tabmove(index = "0") {
     let newindex: number
     if (index.startsWith("+") || index.startsWith("-")) {
         newindex = Math.max(0, Number(index) + aTab.index)
+        if (newindex === aTab.index) return
+        let candidates = await browser.tabs.query({windowId: aTab.windowId})
+        newindex = Math.min(newindex, candidates.length)
+        candidates.sort((a,b) => {
```

Comment:
Sort of heavy-handed, but we get everything in one async query plus a mergesort, if that's still what Mozilla uses. The subsequent slice is straight-forward and we skip the need to filter separately by just iterating over the eligible items and stopping on the first match.

Hope these comments aren't excessive; I actually had this sitting for quite a while and needed to justify to myself that this was a good enough way to do it.

Reply:
These comments are good. A condensed version should appear in the code.

Here's my version:

```ts
    if (index.startsWith("+") || index.startsWith("-")) {
        newindex = aTab.index + toNumber(index)
        // A non-pinned tab may not have an index less than any pinned tab.  So
        // pinned tabs can't move into the pinned area and pinned tabs can't
        // move out of it.

        // Sort tabs in ascending index order.
        const tabs = await browser.tabs.query({windowId: aTab.windowId})
        tabs.sort((a, b) => {
            return a.index - b.index
        })

        // Slice tabs so that our first choice is index 0, next best is index 1, etc.
        let candidates
        if (newIndex < aTab.index) {
            candidates = tabs.slice(newindex, aTab.index)
        } else {
            // This will also handle +0, -0.
            candidates = tabs.slice(aTab.index, newindex + 1).reverse()
        }

        // Select the first candidate that shares our pinned-ness
        for (const tab of candidates) {
            if (aTab.pinned === tab.pinned) {
                newindex = tab.index
                break
            }
        }
    } else {
```